Embroider update (#2215)
mainmatter/mainmatter.com: src/posts/2023-11-06-embroider-initiative-progress.md

The Embroider Initiative aims to:

1. **Finish Embroider** itself by assigning experienced Ember engineers full-time to the project. This covers development work on Embroider and its dependencies, as well as helping backers setting it up in their repos to uncover and fix edge cases.
2. **Make Embroider maintainable** by decentralizing technical knowledge beyond the project's core developer [Ed Falkner](https://github.com/ef4) and thus [improving the bus factor](https://en.wikipedia.org/wiki/Bus_factor). Mainmatter makes use of the apprenticeship model to onboard and train new developers on the intricates parts of the project.
3. **Shift the ecosystem and make Embroider mainstream** by making it easier to generate Embroider-optimized Ember apps and supporting addon developers make their addons compatible with Embroider.

> To reach these goals and keep up the momentum, the project needs your financial support. Please [get in touch] to learn how to directly benefit from this investment.

Author: I don't know what this link is supposed to point to but I'm assuming it should be a link 😂

## Progress for Embroider Initiative backers

For the Embroider Initiative to be successful, it needs backing from companies that would benefit from a modern and thriving Ember ecosystem. But the benefits of sponsoring the project go beyond the end goal.

While all sponsorship tiers (starting at 3k€) include a backer's logo in Mainmatter communication around the Initiative, companies committing 18k€ to the project get access to the weekly 1-hour call with the team and get opportunities to discuss their technical needs and challenges related to Embroider.

Premium tiers of 36k€ and above include a 2-hour weekly private session with Mainmatter engineers focused on improving the backer's Ember build. In practice, these backers have seen a great return on their investment as the Mainmatter Embroider team has been able to deliver dramatic speed improvements to their build, removing a bottleneck in their development process. Some issues specific to these backers also resulted in solutions that the larger Ember community can now benefit from.

> Let's see some success stories in details:

Author: ```suggestion
Let's see some success stories in detail:
```

### Ticketsolve

Like many companies, Ticketsolve has an internal addon that they use throughout their other applications. Internal addons are a great place to start when you are thinking about upgrading to Embroider for two reasons:

1. It's an isolated place to opt into Embroider's stricter requirements
2. There is a [clear specification](https://github.com/embroider-build/embroider/blob/main/docs/spec.md) of what your v2 addon can do.

> Chris work improving the ergonomics of the v2 addon blueprint in the early weeks of the Embroider Initiative, proved very useful during the process of converting Ticketsolve's internal addon.

Author: ```suggestion
Chris' work improving the ergonomics of the v2 addon blueprint in the early weeks of the Embroider Initiative, proved very useful during the process of converting Ticketsolve's internal addon.
```

This is only a fraction of the changes that came out of these pairing sessions, but it gives a flavour of the kinds of things that we have been able to achieve. While our Senior Engineer Chris Manson has been instrumental in helping speed up this work, Ticketsolve's engineer [Andreas Minnich](https://github.com/enspandi) deserves a shout-out for the colossal amount of work they have put into this effort between pairing sessions.

### Intercom

Intercom also had an internal addon they wanted to convert to v2 before working on converting their main application. Mainmatter worked with them to convert that and, incidentally, we also helped them to convert the tests app that was documenting the v2 addon to use Embroider with the  `staticAddonTestSupportTrees` and `staticAddonTrees` flags turned on.

> While this work was going on, Intercom Engineer [Aaron Chambers](https://github.com/achambers) had setup a CI build step that would keep track if the main Intercom app could build with Embroider and if any tests were passing. Because of this CI job, Aaron identified a 10x slowdown in build times between Embroider v3.1 and v3.2. Chris used a number of pairing sessions to dive into the issue and produce a number of flamecharts for their Embroider builds to get an insight into what was causing their specific slowdown. Those flamecharts were pointing at a particular part of Embroider Package Cache not working exactly as we intended it to. Ed Faulkner managed to pinpoint the problem (while we were discussing it in the "hallway track" of [EmberFest](https://emberfest.eu)) and [opened a pull-request with the fix](https://github.com/embroider-build/embroider/pull/1609). This completely fixed the performance regression for Intercom.

Author: ```suggestion
While this work was going on, Intercom Engineer [Aaron Chambers](https://github.com/achambers) had setup a CI build step that would keep track if the main Intercom app could build with Embroider and if any tests were passing. Because of this CI job, Aaron identified a 10x slowdown in build times between Embroider v3.1 and v3.2. Chris used a number of pairing sessions to dive into the issue and produce a number of flamecharts for their Embroider builds to get an insight into what was causing their specific slowdown. Those flamecharts were pointing at a particular part of Embroider Package Cache not working exactly as we intended it to. Ed Faulkner managed to pinpoint the problem (while he and Chris were discussing it in the "hallway track" of [EmberFest](https://emberfest.eu)) and [opened a pull-request with the fix](https://github.com/embroider-build/embroider/pull/1609). This completely fixed the performance regression for Intercom.
```

## Improving the bus factor

The goals of Embroider may seem simple from the outside, i.e. "just use Webpack or Vite to build your Ember app", but when digging a bit deeper, it's easy to see how complex of a project it really is. This complexity arises from the project having challenging design constraints which pose a significant challenge to anyone who would like to contribute to the core of Embroider. The main design constraint that causes a lot of this complexity arises from the fact that the team wants to provide an easy on-ramp for existing Ember apps to convert to Embroider, and then slowly move those apps from full-compatibility mode to a "fully static" build that can automatically benefit from tree-shaking and code-splitting. This means that we need to provide systems that can automatically convert the still-supported conventions of an Ember app to fully standard compliant ESM code. This is a significant challenge since some of the patterns that are still **officially** supported today date back to the 1.x series of Ember which was released in 2015.

> Because of the overall complexity of project and the rapid iteration of the internal architecture, it hasn't been practical to create any documentation that could communicate what is going on under the hood. This is illustrated by the fact that [Aaron Chambers](https://github.com/achambers) and [Peter Meehan](https://github.com/22a) put together a [document trying to explain the full architecture of the project](https://github.com/embroider-build/embroider/pull/1459), only to have be completely outdated less than a month later when Ed Falkner merged [the first of many internal refactors](https://github.com/embroider-build/embroider/pull/1435).

Author: ```suggestion
Because of the overall complexity of project and the rapid iteration of the internal architecture, it hasn't been practical to create any documentation that could communicate what is going on under the hood. This is illustrated by the fact that [Aaron Chambers](https://github.com/achambers) and [Peter Meehan](https://github.com/22a) put together a [document trying to explain the full architecture of the project](https://github.com/embroider-build/embroider/pull/1459), only to have it be completely outdated less than a month later when Ed Falkner merged [the first of many internal refactors](https://github.com/embroider-build/embroider/pull/1435).
```

Creating a new Ember app with the `embroider` flag (`ember new my-super-app --embroider`) generates a "full compat" app were none of the [Embroider optimisation flags](https://github.com/embroider-build/embroider#options) are passed to the build for the developer.

While more existing apps will work with a "full compat" mode, Ember has reached a point where it makes sense for **newly generated apps** to start with a high water mark so that developers don't accidentally or unconsciously add functionality that won't work in a fully optimised Embroider application. Developers have the option to turn off any of the optimisation flags, but it would be a deliberate choice as they would need to add a specific dependency or functionality to their app.

The Mainmatter Embroider team opened a [PR to switch the functionality](https://github.com/ember-cli/ember-cli/pull/10370) so that passing the `--embroider` flag uses Embroider optimised by default. This involved working through issues with some of the slow test suites that rely on a custom package caching mechanism. The PR got merged and the functionality will be part of Ember v5.5.

### `scenario-tester` ESM compatibility

> [Scenario tester](https://github.com/embroider-build/scenario-tester) is a testing tool that the team makes use of a lot when testing Embroider and `ember-auto-import`. It allows us to generate many scenarios with different combinations of dependencies. We tried using it outside of Embroider, or more specifically outside of a Typescript project, and saw it doesn't work in a CJS environment. [Chris started the effort to move the Typescript build to output an ESM-compatible build](https://github.com/embroider-build/scenario-tester/pull/18) so it can be consumed directly in ESM without a build step. The only remaining thing to do is test if it will work in Embroider before merging and releasing the new version.

Author: ```suggestion
[Scenario tester](https://github.com/embroider-build/scenario-tester) is a testing tool that the team makes use of a lot when testing Embroider and `ember-auto-import`. It allows us to generate many scenarios with different combinations of dependencies. We tried using it outside of Embroider, or more specifically outside of a Typescript project, and saw it doesn't work in a CJS environment. [Chris started the effort to move the Typescript build to output an ESM-compatible build](https://github.com/embroider-build/scenario-tester/pull/18) so it can be consumed directly in ESM without a build step. This change has been tested against the Embroider CI and was recently released in [v3.0.0](https://github.com/embroider-build/scenario-tester/releases/tag/v3.0.0).
```

  
### Documenting the scenario-tester library.

> [`scenario-tester`](https://github.com/ef4/scenario-tester/) is to Embroider what `ember-try` is to Ember CLI: it's a tool that lets us perform automated tests with various combinations of dependencies, configs and circumstances. The approach of `scenario-tester` is different: instead of reinstalling dependencies for every test case, it has all dependencies (including all versions of dependencies) set up once, saving a lot of time. It leverages [fixturify-project](https://github.com/stefanpenner/node-fixturify-project) to create and emit to filesystem Ember apps and addons with predefined dependencies and configuration, in order to run tests on them.

Author: We should add  a concluding line here that says something like "this means it's quite a complex project and can be challenging to use. As we are making Embroider more mainstream we want people to be able to contribute failing tests for bugs that the discover so it's very important to have the system that we rely on for testing be more easily accessible by the community"

### Working on the reverse-exports

During the build, Embroider needs to expose Ember internals to Vite and Webpack in a way they can understand and consume. Modern Ember apps can have multiple [exports](https://nodejs.org/api/packages.html#package-entry-points) entry points in their `package.json` configs. This poses a peculiar challenge for Embroider: it needs to know how to reorganize files in an Ember project in such a way that they would resolve into paths defined as `exports` values. Essentially, this requires resolving `exports` in reverse, and this it what the [reverse-exports](https://github.com/embroider-build/embroider/pull/1652/) package is for.

## Call to action

> We are hoping we can extend the initiative's budget and timeline to keep up the momentum in 2024. Please [get in touch] to learn how to financially support the Embroider Initiative and directly benefit from this investment.

Author: again we need the get in touch to be a link of some sort 😂
